source/postcard-c/postcard.h

```diff
@@ -0,0 +1,1184 @@
+/// - value - the value to encode
+///
+/// *Side Effects / Return*
+/// If there is not enough room in the buffer `postcard_error_t` will return a
+/// non-zero value If encoding was successful, slice.len will be incremented by
+/// the number of encoded bytes
+postcard_error_t postcard_encode_i32(postcard_slice_t* slice, int32_t value);
+/// Encodes a u64 to the passed in slice.
+///
+/// *Arguments*
+/// - slice - a pointer to an initialized `postcard_slice_t`
+/// - value - the value to encode
+///
+/// *Side Effects / Return*
+/// If there is not enough room in the buffer `postcard_error_t` will return a
+/// non-zero value If encoding was successful, slice.len will be incremented by
+/// the number of encoded bytes
+postcard_error_t postcard_encode_u64(postcard_slice_t* slice, uint64_t value);
+/// Encodes a i64 to the passed in slice.
+///
+/// *Arguments*
+/// - slice - a pointer to an initialized `postcard_slice_t`
+/// - value - the value to encode
+///
+/// *Side Effects / Return*
+/// If there is not enough room in the buffer `postcard_error_t` will return a
+/// non-zero value If encoding was successful, slice.len will be incremented by
+/// the number of encoded bytes
+postcard_error_t postcard_encode_i64(postcard_slice_t* slice, int64_t value);
+/// Encodes a f32 to the passed in slice.
+///
+/// *Arguments*
+/// - slice - a pointer to an initialized `postcard_slice_t`
+/// - value - the value to encode
+///
+/// *Side Effects / Return*
+/// If there is not enough room in the buffer `postcard_error_t` will return a
+/// non-zero value If encoding was successful, slice.len will be incremented by
+/// the number of encoded bytes
+postcard_error_t postcard_encode_f32(postcard_slice_t* slice, float value);
+
+/// Encodes a f64 to the passed in slice.
+///
+/// *Arguments*
+/// - slice - a pointer to an initialized `postcard_slice_t`
+/// - value - the value to encode
+///
+/// *Side Effects / Return*
+/// If there is not enough room in the buffer `postcard_error_t` will return a
+/// non-zero value If encoding was successful, slice.len will be incremented by
+/// the number of encoded bytes
+postcard_error_t postcard_encode_f64(postcard_slice_t* slice, double value);
+
+/// Encodes a byte array to the passed in slice.
+///
+/// *Arguments*
+/// - slice - a pointer to an initialized `postcard_slice_t`
+/// - bytes - a pointer to the byte array to encode
+/// - length - the length of the byte array
+/// *Safety*
+/// The user must ensure that bytes points to a valid initialized memory, and
+/// that length is not greater than the size of the array.
+///
+/// *Side Effects /
+/// Returns* If there is not enough room in the buffer `postcard_error_t` will
+/// return a non-zero value If encoding was successful, slice.len will be
+/// incremented by the number of encoded bytes
+postcard_error_t postcard_encode_byte_array(postcard_slice_t* slice,
+    const uint8_t* bytes,
+    size_t length);
```

Comment:
Would it be helpful to have a way to determine if there is enough room in the slice to push the array, without actually writing anything? (And other functions like this)

Reply:
This is a potential problem across all of the functions, not just byte array. The trouble is that it would make encoding essentially a two-step operation: one to check how much size a varint takes, and another to actually encode it. 

In the use cases I envisioned for this, I don't think this will cause practical problems. The first is a higher-level code generator that will pre-allocate memory based on the calculated encoded size. Another is serializing messages to fixed-size buffers. In that case, I'm not too concerned if there is half-written data, as long as the functions properly error out.

I think we can recommend that users run the postcard_size functions themselves if they never want to write data when the buffer isn't big enough.

Reply:
a bit tangential to the discussion, but relevant for this (and similar APIs):
rather than passing a pointer you could pass an array - there's a fancy C notation (since C99) which allows you to specify the parameter which gives the length and the compiler will validate it. but the length parameter has to come before the array, otherwise it doesn't work (until [this proposal](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3394.pdf) gets standardised).

this works:
```c
int foo(unsigned int len, char foo[static len]);
```